sync/src/types/mod.rs

```diff
@@ -69,13 +69,60 @@ const SHRINK_THRESHOLD: usize = 300;
 //     If their best known block is still behind when that new timeout is
 //     reached, disconnect.
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 pub struct ChainSyncState {
     pub timeout: u64,
     pub work_header: Option<core::HeaderView>,
     pub total_difficulty: Option<U256>,
     pub sent_getheaders: bool,
-    pub not_sync_until: Option<u64>,
+    headers_sync_state: HeadersSyncState,
+}
+
+impl Default for ChainSyncState {
+    fn default() -> Self {
+        Self {
+            timeout: 0,
+            work_header: None,
+            total_difficulty: None,
+            sent_getheaders: false,
+            headers_sync_state: HeadersSyncState::Initialized,
+        }
+    }
+}
+
+impl ChainSyncState {
+    fn can_sync(&self, now: u64) -> bool {
+        match self.headers_sync_state {
+            HeadersSyncState::Initialized => false,
+            HeadersSyncState::SyncProtocolConnected => true,
+            HeadersSyncState::Started => false,
+            HeadersSyncState::Suspend(timestamp) => timestamp < now,
+        }
+    }
+
+    fn connected(&mut self) {
+        self.headers_sync_state = HeadersSyncState::SyncProtocolConnected;
+    }
+
+    fn start(&mut self) {
+        self.headers_sync_state = HeadersSyncState::Started
+    }
+
+    fn suspend(&mut self, timestamp: u64) {
+        self.headers_sync_state = HeadersSyncState::Suspend(timestamp)
+    }
+
+    fn started(&self) -> bool {
+        matches!(self.headers_sync_state, HeadersSyncState::Started)
+    }
+}
+
+#[derive(Clone, Debug)]
+enum HeadersSyncState {
+    Initialized,
+    SyncProtocolConnected,
+    Started,
+    Suspend(u64),
```

Comment:
Better to add some comment about the `u64` field. What does it mean and what unit it uses.

Reply:
@doitian updated accordingly, please review again, thanks.